plugin/generator/templates.go

```diff
@@ -0,0 +1,163 @@
+			// TODO: only a problem if we need to deal with _1...
+			return strings.TrimPrefix(service.Spec.Name, "/"), nil
+		},
+		"labels": func(options ...interface{}) (map[string]string, error) {
+			// TODO: mix in image labels...
+			return service.Spec.Labels, nil
+		},
+		"hostname": func(options ...interface{}) (string, error) {
+			// if there is a string param, use it.
+			if len(options) == 1 {
+				if host, isString := options[0].(string); isString && host != "" {
+					return host, nil
+				}
+			}
+			// TODO: how do we deal if we have a full domain name?
+
+			// TODO: from compose, looks like caddy-docker-proxy_maintainence_1 (remove _1?)
+			return strings.TrimPrefix(service.Spec.Name, "/"), nil
+		},
+	}
+	return g.getTemplatedCaddyfile(service, funcMap, logsBuffer)
+}
+
+func (g *CaddyfileGenerator) getContainerTemplatedCaddyfile(container *types.Container, logsBuffer *bytes.Buffer) (*caddyfile.Container, error) {
+
+	logsBuffer.WriteString(fmt.Sprintf("[INFO] getTemplatedCaddyfile \n"))
+
+	funcMap := template.FuncMap{
+		"entitytype": func(options ...interface{}) (string, error) {
+			return "container", nil
+		},
+		"upstreams": func(options ...interface{}) (string, error) {
+			targets, err := g.getContainerIPAddresses(container, logsBuffer, true)
+			transformed := []string{}
+			for _, target := range targets {
+				for _, param := range options {
+					if protocol, isProtocol := param.(string); isProtocol {
+						target = protocol + "://" + target
+					} else if port, isPort := param.(int); isPort {
+						target = target + ":" + strconv.Itoa(port)
+					}
+				}
+				transformed = append(transformed, target)
+			}
+			return strings.Join(transformed, " "), err
+		},
+		"matcher": func(options ...interface{}) (string, error) {
+			// TODO: only a problem if we need to deal with _1...
+			return strings.TrimPrefix(container.Names[0], "/"), nil
+		},
+		"labels": func(options ...interface{}) (map[string]string, error) {
+			// TODO: mix in image labels...
+			return container.Labels, nil
+		},
+		"hostname": func(options ...interface{}) (string, error) {
+			// if there is a string param, use it.
+			if len(options) == 1 {
+				if host, isString := options[0].(string); isString && host != "" {
+					return host, nil
+				}
+			}
+			// TODO: how do we deal if we have a full domain name?
+
+			// TODO: from compose, looks like caddy-docker-proxy_maintainence_1 (remove _1?)
+			return strings.TrimPrefix(container.Names[0], "/"), nil
+		},
+	}
+	return g.getTemplatedCaddyfile(container, funcMap, logsBuffer)
+}
+
+func (g *CaddyfileGenerator) getTemplatedCaddyfile(data interface{}, funcMap template.FuncMap, logsBuffer *bytes.Buffer) (*caddyfile.Container, error) {
+
+	commonFuncMap := template.FuncMap{
+		"http": func() string {
+			return "http"
+		},
+		"https": func() string {
+			return "https"
+		},
+	}
+
+	// TODO: how to deal with _1, _2 etc - multiple routes on one container...
```

Comment:
Would it be possible to handle it at template level by doing something like:
```
for i = 0; i < 10; i++ {
  hostLabel := "virtual_" + i + ".host";
  portLabel := "virtual_" + i + ".port";
  {{ if index labels hostLabel }}
  ....
  {{ end }}
}
```
Don't know how to write a proper template to do that, but I think you will get the idea from the sketch above

Reply:
To do the for loop, you can use range and one of the following sprig functions:
http://masterminds.github.io/sprig/integer_slice.html

Reply:
@lucaslorentz I'm wondering if you want the `virtual.port` etc labels to be in your plugin, or if you'd like them to be examples of what a user _could_ use the template generator for.

Reply:
I think they should be only template examples.